garak/attempt.py

```diff
@@ -13,21 +16,153 @@
+    they may also be (or include) images, audio, files, or even a composition of
+    these. The Turn object encapsulates this flexibility.
+    `Message` doesn't yet support multiple attachments of the same type.
+
+    :param text: Text of the prompt/response
+    :type text: str
+    :param data_path: Path to attachment
+    :type data_path: Union[str, Path]
+    :param data_type: (type, encoding) Mime type of data
+    :type data_type: Tuple(str, str)
+    :param data_checksum: sha256 checksum of data loaded
+    :type data_checksum: str
+    :param data: Data to attach
+    :type data: Any
+    :param lang: language code for `text` content
+    :type lang: str (bcp47 language code)
+    :param notes: Free form dictionary of notes for the turn
+    :type notes: dict
+    """
+
+    text: str = None
+    lang: str = None
+    data_path: Optional[str] = None
+    data_type: Optional[str] = None
+    data_checksum: Optional[str] = None
+    # data: bytes = None  # should this dataclass attribute exist?
+    notes: Optional[dict] = field(
+        default_factory=dict
+    )  # is this valid for a dataclass?
+
+    @property
+    def data(self):
+        if not hasattr(self, "_data"):
+            if self.data_path is not None:
+                import hashlib
+                import mimetypes
+
+                self.data_type = mimetypes.guess_type(self.data_path)
+                self._data = Message._load_data(self.data_path)
+                self.data_checksum = hashlib.sha256(
+                    self._data, usedforsecurity=False
+                ).hexdigest()
+            else:
+                return None
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        if self.data_path is not None and hasattr(self, "_data"):
+            raise ValueError("data_path has been set data cannot be modified")
+        if self.data_type is None:
+            raise ValueError("data_type must be set before data")
+        self._data = value
+        self.data_checksum = None
+        if self._data:
+            import hashlib
+
+            self.data_checksum = hashlib.sha256(
+                self._data, usedforsecurity=False
+            ).hexdigest()
+
+    @staticmethod
+    def _load_data(data_path: Union[str, Path]):
+        with open(data_path, "rb") as f:
+            return f.read()
+
+
+@dataclass
+class Turn:
+    """Object to attach actor context to a message, denoted as taking a `Turn` in the conversation
+
+    :param role: Role of the participant who issued the utterance Expected: ["system", "user", "assistant"]
+    :type role: str
+    """
+
+    role: str
+    content: Message
+
+    @staticmethod
+    def from_dict(value: dict):
+        entity = deepcopy(value)
+        message = entity.pop("content", {})
+        entity["content"] = Message(**message)
+        ret_val = Turn(**entity)
+        return ret_val
+
+
+@dataclass
+class Conversation:
```

Comment:
it seems like the intent is to model multi-turn garak interactions, i.e. conversations, as a linked list of `Attempts`. this is a result of the move in this PR toward making `Attempt`s immutable.

consider then renaming `Conversation`, here, to e.g. `Prompt`. it seems to represent the input posed to a generator, which we commonly refer to as a prompt, regardless of how it is composed or represented in the calling process.

there may be a better alternative than `Prompt` for this, but it doesn't represent a conversation - that's represented in a chain of attempts. this class /does/ represent a turn history. `TurnList` I guess might fit.

Reply:
I think this misreads the _intent_, while this revision does expect to eventually _document_ a multi-turn attack as a series of linked `Attempt` objects each attempt contains the `Conversation` up to that `Attempt` and to enable stochastic testing encapsulates how and when `Conversations` diverge.

This does not make `Attempt` immutable as yet, it strives to ensure the `prompt` the attempt is in reference to is immutable and represent a single round trip through a generator.

Reply:
I will create a diagram or psuedo example